src/lore/lore-calculator.cpp

```diff
@@ -125,20 +126,21 @@ bool know_damage(MonsterRaceId r_idx, int i)
 void set_damage(PlayerType *player_ptr, lore_type *lore_ptr, MonsterAbilityType ms_type, concptr msg)
 {
     MonsterRaceId r_idx = lore_ptr->r_idx;
-    int base_damage = monspell_race_damage(player_ptr, ms_type, r_idx, BASE_DAM);
-    int dice_num = monspell_race_damage(player_ptr, ms_type, r_idx, DICE_NUM);
-    int dice_side = monspell_race_damage(player_ptr, ms_type, r_idx, DICE_SIDE);
-    int dice_mult = monspell_race_damage(player_ptr, ms_type, r_idx, DICE_MULT);
-    int dice_div = monspell_race_damage(player_ptr, ms_type, r_idx, DICE_DIV);
-    char dmg_str[80], dice_str[sizeof(dmg_str) + 10];
     char *tmp = lore_ptr->tmp_msg[lore_ptr->vn];
-    dice_to_string(base_damage, dice_num, dice_side, dice_mult, dice_div, dmg_str);
-    sprintf(dice_str, "(%s)", dmg_str);
+    size_t tmpsz = sizeof(lore_ptr->tmp_msg[lore_ptr->vn]);
 
     if (know_armour(r_idx, lore_ptr->know_everything)) {
-        sprintf(tmp, msg, dice_str);
+        int base_damage = monspell_race_damage(player_ptr, ms_type, r_idx, BASE_DAM);
+        int dice_num = monspell_race_damage(player_ptr, ms_type, r_idx, DICE_NUM);
+        int dice_side = monspell_race_damage(player_ptr, ms_type, r_idx, DICE_SIDE);
+        int dice_mult = monspell_race_damage(player_ptr, ms_type, r_idx, DICE_MULT);
+        int dice_div = monspell_race_damage(player_ptr, ms_type, r_idx, DICE_DIV);
+        char dmg_str[80] = "(";
+        dice_to_string(base_damage, dice_num, dice_side, dice_mult, dice_div, dmg_str + 1);
+        angband_strcat(dmg_str, ")", sizeof(dmg_str));
+        strnfmt(tmp, tmpsz, msg, dmg_str);
     } else {
-        sprintf(tmp, msg, "");
+        tmp[0] = '\0';
```

Comment:
It would be better to bring the condition of the else clause first and return early.

```c++
if (!know_armour(...)) {
    tmp[0] = '\0';
    return;
}

...
```

Reply:
The forced push puts the else clause there first and has it return early.

Reply:
I checked this section again, and it looks like the argument `msg` contains a string like `_("酸%s", "acid%s")` and `sprintf(tmp, msg, "")` in the else clause means that the %s part is empty.
Therefore, instead of `tmp[0] = '\0';`, wouldn't it be correct to use `strnfmt(tmp, tmpsz, msg, "");`?
Please confirm.

Reply:
That's correct.  The latest forced push should address that.  It also adds the changes to player-info/self-info.cpp from #2861 since they fit the criteria for this pull request.